File: Assets/Disease Destroyer Assets/Scripts/cellMovement.js
var splat : AudioClip;//death sound
var oneShotAudio=false;//singleton for death sound
private var maxX = 480;//bounding box
private var minX = -480;
private var maxY = 480;
private var minY = -480;
private var health=3;
private var emitter: ParticleEmitter;
public var tex1: Texture2D;
public var tex2: Texture2D;
public var tex3: Texture2D;
public var tex4: Texture2D;
private var startTime;
var virus : Rigidbody;
// Use this for initialization
var virusClone: Transform;
function Start () {
	emitter = GetComponentInChildren(ParticleEmitter);   
    if (emitter) {
        emitter.emit = false;
    }
    this.startTime=Time.time;
}
 
// Update is called once per frame
function Update () {
	if(this.startTime==null){
		this.startTime=Time.time;
	}
	//if(rigidbody.velocity.magnitude>=50){
	//	rigidbody.velocity=rigidbody.velocity*.95;
	//}
	if(rigidbody.transform.position.x>maxX){//if the virus leaves the bounding box bring it back in and send it toward the center
    	rigidbody.transform.position.x=maxX;
    	rigidbody.velocity.x=-Mathf.Abs(rigidbody.velocity.x);
    }
    if(rigidbody.transform.position.x<minX){
    	rigidbody.transform.position.x=minX;
    	rigidbody.velocity.x=Mathf.Abs(rigidbody.velocity.x);
    }
    if(rigidbody.transform.position.y>maxY){
    	rigidbody.transform.position.y=maxY;
    	rigidbody.velocity.y=-Mathf.Abs(rigidbody.velocity.y);
    }
    if(rigidbody.transform.position.y<minY){
    	rigidbody.transform.position.y=minY;
    	rigidbody.velocity.y=Mathf.Abs(rigidbody.velocity.y);
    }
}

function OnCollisionEnter(collision : Collision) {
	//var tex = Resources.Load("cell3");
	
	if(collision.gameObject.name.Contains("Border")){//bounce off borders
		if(rigidbody.position.x>0&&Mathf.Abs(rigidbody.position.x)>450){
			rigidbody.velocity.x=-Mathf.Abs(rigidbody.velocity.x);
		}
		if(rigidbody.position.x<0&&Mathf.Abs(rigidbody.position.x)>450){
			rigidbody.velocity.x=Mathf.Abs(rigidbody.velocity.x);
		}
		if(rigidbody.position.y>0&&Mathf.Abs(rigidbody.position.y)>450){
			rigidbody.velocity.y=-Mathf.Abs(rigidbody.velocity.y);
		}
		if(rigidbody.position.y<0&&Mathf.Abs(rigidbody.position.y)>450){
			rigidbody.velocity.y=Mathf.Abs(rigidbody.velocity.y);
		}
    }
    if(this.startTime==null){
		this.startTime=Time.time;
	}/*
	if(collision.gameObject.name.Contains("Character")){
		renderer.material.mainTexture = tex3;
	}*/
	if(collision.gameObject.name.Contains("Virus")&&(Time.time>(this.startTime+2))){//destroy 2 seconds after a viruses touch
		kill(false);
    }
 
}
function corrupt(){
	renderer.material.mainTexture= tex4;
	if(emitter!=null){
		emitter.emit=true;
		emitter.transform.parent=null; // detach particle system
		Destroy(emitter.gameObject, 3);
	}
	if (splat) AudioSource.PlayClipAtPoint(splat, transform.position);
	spawnVirus();
	Destroy (gameObject);
	Destroy (this);
	Destroy (rigidbody);
}
function kill(spawn: boolean){
	health=health-1;
		if(health==2){renderer.material.mainTexture = tex2;}
		if(health==1){renderer.material.mainTexture= tex3;}
		if(health==0){
			
			if(emitter!=null){
				emitter.emit=true;
				emitter.transform.parent=null; // detach particle system
				Destroy(emitter.gameObject, 3);
			}
			//if(!oneShotAudio){//play death sound and destroy
			if (splat) AudioSource.PlayClipAtPoint(splat, transform.position);
				//oneShotAudio=true;
			//}
			if(spawn)spawnVirus();
			//virusClone=Instantiate(virus, Vector3(rigidbody.x,rigidbody.y,rigidbody.z));
    		Destroy (gameObject);
			Destroy (this);
			Destroy (rigidbody);
			
		}
		this.startTime=Time.time;
}
function spawnVirus(){
	var newx=rigidbody.position.x;
	var newy=rigidbody.position.y;
	var virusClone : Rigidbody = Instantiate(virus, Vector3(newx,newy,0), virusClone.rotation);
	virusClone.transform.parent=null;
}

File: Assets/Disease Destroyer Assets/Scripts/cellMovement.cs
using UnityEngine;

[RequireComponent(typeof(Rigidbody),typeof(Renderer))]
public class cellMovement : MonoBehaviour
{
    public AudioClip splat; public bool oneShotAudio; public Texture2D tex1,tex2,tex3,tex4; public Rigidbody virus; public Transform virusClone; public ParticleSystem emitter;
    int health=3; float startTime; Rigidbody body; Renderer visual;
    void Awake(){body=GetComponent<Rigidbody>();visual=GetComponent<Renderer>();emitter=GetComponentInChildren<ParticleSystem>();}
    void Start(){startTime=Time.time;if(emitter)emitter.Stop();}
    void FixedUpdate(){if(Time.timeScale>0)Bounce();}
    void OnCollisionEnter(Collision c){if(c.gameObject.name.Contains("Virus")&&Time.time>startTime+2)kill(false);}
    void OnCollisionStay(Collision c){if(!c.gameObject.CompareTag("Respawn")||!c.rigidbody)return;var delta=body.position-c.rigidbody.position;delta.z=0;var distance=delta.magnitude;if(distance>=40f)return;var direction=distance>.001f?delta/distance:Vector3.right;var correction=direction*((40f-distance)*.5f);body.position+=correction;c.rigidbody.position-=correction;}
    void Bounce(){var p=body.position;var v=body.linearVelocity;if(p.x>480){p.x=480;v.x=-Mathf.Abs(v.x);}if(p.x<-480){p.x=-480;v.x=Mathf.Abs(v.x);}if(p.y>480){p.y=480;v.y=-Mathf.Abs(v.y);}if(p.y<-480){p.y=-480;v.y=Mathf.Abs(v.y);}body.position=p;body.linearVelocity=v;}
    public void corrupt(){if(visual&&tex4)visual.material.mainTexture=tex4;Emit(3);HUDHandler.Explode(transform.position,new Color(.85f,.2f,.05f),1.1f);HUDHandler.PlaySfx(splat);SpawnVirus();Destroy(gameObject);}
    public void kill(bool spawn){health--;if(visual){if(health==2&&tex2)visual.material.mainTexture=tex2;if(health==1&&tex3)visual.material.mainTexture=tex3;}if(health<=0){Emit(3);HUDHandler.Explode(transform.position,new Color(1f,.25f,.05f),1.1f);HUDHandler.PlaySfx(splat);if(spawn)SpawnVirus();Destroy(gameObject);}startTime=Time.time;}
    void SpawnVirus(){if(virus)Instantiate(virus,new Vector3(body.position.x,body.position.y,0),virus.rotation);}
    void Emit(float life){if(!emitter)return;emitter.transform.SetParent(null);emitter.Play();Destroy(emitter.gameObject,life);}
}
